pkg/quota/admission/clusterresourceoverride/admission.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		glog.V(5).Infof("%s error reading config: %v", api.PluginName, err)
+		return nil, err
+	}
+	glog.V(5).Infof("%s config is: %v", api.PluginName, config)
+	return config, nil
+}
+
+func (a *clusterResourceOverridePlugin) Validate() error {
+	if errs := validation.Validate(a.Config); len(errs) > 0 {
+		return errs.ToAggregate()
+	}
+	if a.ProjectCache == nil {
+		return fmt.Errorf("%s did not get a project cache", api.PluginName)
+	}
+	return nil
+}
+
+// TODO this will need to update when we have pod requests/limits
+func (a *clusterResourceOverridePlugin) Admit(attr admission.Attributes) error {
+	glog.V(6).Infof("%s admission controller is invoked", api.PluginName)
+	if a.Config == nil || attr.GetResource() != kapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil // not applicable
+	}
+	pod, ok := attr.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+	glog.V(5).Infof("%s is looking at creating pod %s in project %s", api.PluginName, pod.Name, attr.GetNamespace())
+
+	// allow annotations on project to override
+	if ns, err := a.ProjectCache.GetNamespace(attr.GetNamespace()); err != nil {
+		glog.Warningf("%s got an error retrieving namespace: %v", api.PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	} else {
+		projectEnabledPlugin, exists := ns.Annotations[clusterResourceOverrideAnnotation]
+		if exists && projectEnabledPlugin != "true" {
+			glog.V(5).Infof("%s is disabled for project %s", api.PluginName, attr.GetNamespace())
+			return nil // disabled for this project, do nothing
+		}
+	}
+
+	// Reuse LimitRanger logic to apply limit/req defaults from the project. Ignore validation
+	// errors, assume that LimitRanger will run after this plugin to validate.
+	glog.V(5).Infof("%s: initial pod limits are: %#v", api.PluginName, pod.Spec.Containers[0].Resources)
+	if err := a.LimitRanger.Admit(attr); err != nil {
+		glog.V(5).Infof("%s: error from LimitRanger: %#v", api.PluginName, err)
+	}
+	glog.V(5).Infof("%s: pod limits after LimitRanger are: %#v", api.PluginName, pod.Spec.Containers[0].Resources)
+	for _, container := range pod.Spec.Containers {
+		resources := container.Resources
+		memLimit, memFound := resources.Limits[kapi.ResourceMemory]
+		if memFound && a.Config.MemoryRequestToLimitPercent != 0 {
+			resources.Requests[kapi.ResourceMemory] = resource.Quantity{
+				Amount: multiply(memLimit.Amount, a.internalConfig.memoryRequestToLimitRatio),
+				Format: resource.BinarySI,
+			}
+		}
+		if memFound && a.Config.LimitCPUToMemoryPercent != 0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
+				// float math is necessary here as there is no way to create an inf.Dec to represent cpuBaseScaleFactor < 0.001
+				Amount: multiply(inf.NewDec(int64(float64(memLimit.Value())*cpuBaseScaleFactor), 3), a.internalConfig.limitCPUToMemoryRatio),
```

Comment:
>
> I don't see a better way.  If you have a good test that covers large memory values for what is realistically possible, then I think we are fine.
>
>
I added some tests for large and small values. There is some loss of
precision when you get down below millicores but I think that's unavoidable
with this library, and not significant to users.